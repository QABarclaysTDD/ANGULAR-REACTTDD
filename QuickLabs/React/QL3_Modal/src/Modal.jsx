import React, { useEffect, useRef } from 'react';

const Modal = ({ isOpen, onClose, children }) => {
  const modalRef = useRef(null);
  const modalId = 'modal-title';

  useEffect(() => {
    if (isOpen && modalRef.current) {
      // Focus the modal when it opens
      modalRef.current.focus();
    }
  }, [isOpen]);

  useEffect(() => {
    const handleEscape = (event) => {
      if (event.key === 'Escape' && isOpen) {
        onClose();
      }
    };

    if (isOpen) {
      document.addEventListener('keydown', handleEscape);
      // Prevent body scroll when modal is open
      document.body.style.overflow = 'hidden';
    }

    return () => {
      document.removeEventListener('keydown', handleEscape);
      document.body.style.overflow = 'unset';
    };
  }, [isOpen, onClose]);

  if (!isOpen) {
    return null;
  }

  const handleOverlayClick = (event) => {
    if (event.target === event.currentTarget) {
      onClose();
    }
  };

  const handleContentClick = (event) => {
    // Prevent closing when clicking inside modal content
    event.stopPropagation();
  };

  return (
    <div
      data-testid="modal-overlay"
      style={{
        position: 'fixed',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        backgroundColor: 'rgba(0, 0, 0, 0.5)',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        zIndex: 1000,
      }}
      onClick={handleOverlayClick}
    >
      <div
        ref={modalRef}
        role="dialog"
        aria-modal="true"
        aria-labelledby={modalId}
        tabIndex={-1}
        style={{
          backgroundColor: 'white',
          borderRadius: '8px',
          padding: '24px',
          maxWidth: '90vw',
          maxHeight: '90vh',
          overflow: 'auto',
          position: 'relative',
          outline: 'none',
        }}
        onClick={handleContentClick}
      >
        <button
          type="button"
          onClick={onClose}
          aria-label="Close modal"
          style={{
            position: 'absolute',
            top: '8px',
            right: '8px',
            background: 'none',
            border: 'none',
            fontSize: '24px',
            cursor: 'pointer',
            padding: '8px',
            lineHeight: 1,
            color: '#666',
          }}
        >
          Ã—
        </button>
        <div id={modalId}>
          {children}
        </div>
      </div>
    </div>
  );
};

export default Modal;