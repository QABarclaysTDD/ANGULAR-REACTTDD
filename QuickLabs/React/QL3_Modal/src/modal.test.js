import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Modal from './Modal';

describe('Modal Component', () => {
  const defaultProps = {
    isOpen: false,
    onClose: jest.fn(),
    children: <div>Modal Content</div>,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Snapshot Testing', () => {
    test('should match snapshot when modal is closed', () => {
      const { asFragment } = render(
        <Modal {...defaultProps} isOpen={false}>
          <div>Modal Content</div>
        </Modal>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    test('should match snapshot when modal is open', () => {
      const { asFragment } = render(
        <Modal {...defaultProps} isOpen={true}>
          <div>Modal Content</div>
        </Modal>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('Modal Visibility', () => {
    test('should be hidden by default when isOpen is false', () => {
      render(
        <Modal {...defaultProps} isOpen={false}>
          <div>Modal Content</div>
        </Modal>
      );
      
      expect(screen.queryByText('Modal Content')).not.toBeInTheDocument();
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });

    test('should display modal when isOpen is true', () => {
      render(
        <Modal {...defaultProps} isOpen={true}>
          <div>Modal Content</div>
        </Modal>
      );
      
      expect(screen.getByText('Modal Content')).toBeInTheDocument();
      expect(screen.getByRole('dialog')).toBeInTheDocument();
    });
  });

  describe('Modal Interactions', () => {
    test('should call onClose when close button is clicked', () => {
      const mockOnClose = jest.fn();
      
      render(
        <Modal {...defaultProps} isOpen={true} onClose={mockOnClose}>
          <div>Modal Content</div>
        </Modal>
      );
      
      const closeButton = screen.getByRole('button', { name: /close/i });
      fireEvent.click(closeButton);
      
      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    test('should call onClose when overlay is clicked', () => {
      const mockOnClose = jest.fn();
      
      render(
        <Modal {...defaultProps} isOpen={true} onClose={mockOnClose}>
          <div>Modal Content</div>
        </Modal>
      );
      
      const overlay = screen.getByTestId('modal-overlay');
      fireEvent.click(overlay);
      
      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });

    test('should not call onClose when modal content is clicked', () => {
      const mockOnClose = jest.fn();
      
      render(
        <Modal {...defaultProps} isOpen={true} onClose={mockOnClose}>
          <div>Modal Content</div>
        </Modal>
      );
      
      const modalContent = screen.getByText('Modal Content');
      fireEvent.click(modalContent);
      
      expect(mockOnClose).not.toHaveBeenCalled();
    });
  });

  describe('Accessibility', () => {
    test('should have correct accessibility attributes when open', () => {
      render(
        <Modal {...defaultProps} isOpen={true}>
          <div>Modal Content</div>
        </Modal>
      );
      
      const dialog = screen.getByRole('dialog');
      expect(dialog).toHaveAttribute('aria-modal', 'true');
      expect(dialog).toHaveAttribute('aria-labelledby');
    });

    test('should focus the modal when opened', () => {
      render(
        <Modal {...defaultProps} isOpen={true}>
          <div>Modal Content</div>
        </Modal>
      );
      
      const dialog = screen.getByRole('dialog');
      expect(dialog).toHaveFocus();
    });
  });

  describe('Complete Modal with Open/Close Flow', () => {
    test('should open modal when open button is clicked and close when close button is clicked', () => {
      const TestComponent = () => {
        const [isOpen, setIsOpen] = React.useState(false);
        
        return (
          <div>
            <button onClick={() => setIsOpen(true)}>Open Modal</button>
            <Modal isOpen={isOpen} onClose={() => setIsOpen(false)}>
              <h2>Test Modal</h2>
              <p>This is modal content</p>
            </Modal>
          </div>
        );
      };

      render(<TestComponent />);
      
      // Initially modal should be closed
      expect(screen.queryByText('Test Modal')).not.toBeInTheDocument();
      
      // Click open button
      const openButton = screen.getByText('Open Modal');
      fireEvent.click(openButton);
      
      // Modal should now be visible
      expect(screen.getByText('Test Modal')).toBeInTheDocument();
      expect(screen.getByText('This is modal content')).toBeInTheDocument();
      
      // Click close button
      const closeButton = screen.getByRole('button', { name: /close/i });
      fireEvent.click(closeButton);
      
      // Modal should be hidden again
      expect(screen.queryByText('Test Modal')).not.toBeInTheDocument();
    });
  });
});