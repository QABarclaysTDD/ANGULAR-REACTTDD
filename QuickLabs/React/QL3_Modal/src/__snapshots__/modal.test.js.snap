// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Modal Component Snapshot Testing should match snapshot when modal is closed 1`] = `<DocumentFragment />`;

exports[`Modal Component Snapshot Testing should match snapshot when modal is open 1`] = `
<DocumentFragment>
  <div
    data-testid="modal-overlay"
    style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(0, 0, 0, 0.5); display: flex; align-items: center; justify-content: center; z-index: 1000;"
  >
    <div
      aria-labelledby="modal-title"
      aria-modal="true"
      role="dialog"
      style="background-color: white; border-radius: 8px; padding: 24px; max-width: 90vw; max-height: 90vh; overflow: auto; position: relative; outline: none;"
      tabindex="-1"
    >
      <button
        aria-label="Close modal"
        style="position: absolute; top: 8px; right: 8px; background: none; font-size: 24px; cursor: pointer; padding: 8px; line-height: 1; color: rgb(102, 102, 102);"
        type="button"
      >
        Ã—
      </button>
      <div
        id="modal-title"
      >
        <div>
          Modal Content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
